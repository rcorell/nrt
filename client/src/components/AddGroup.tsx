import { useMutation } from '@apollo/client';
import React, { useState } from 'react';
import { Button, Form, FormControlProps } from 'react-bootstrap';

import { createGroupMutationString } from 'src/api/api';
import { CreateGroupMutation, CreateGroupMutationVariables } from 'src/api/__generated__/CreateGroupMutation';
import { setBrowserTitle } from 'src/utils';
import { AppError, AppForm, FormContainer } from 'src/styles/form';

export const AddGroup: React.FC = () => {
    const [description, setDescription] = useState('');
    const [name, setName] = useState('');

    setBrowserTitle('Add Group');

    const [createGroup, { error }] = useMutation<CreateGroupMutation, CreateGroupMutationVariables>(
        createGroupMutationString,
        {
            variables: { description, name },
            onCompleted: ({ createGroup }) => {
                createGroup.id += 0;
            }
        }
    );

    const isFormInvalid = () => {
        return name.length < 1;
    };

    const handleSubmit = (event: React.FormEvent<FormControlProps>) => {
        event.preventDefault();

        createGroup();
    };

    const errorDisplay = () => {
        if (error) {
            return <div>Error: {JSON.stringify(error)}</div>;
        }

        return null;
    };

    return (
        <FormContainer>
            <h1>Add Group</h1>
            {errorDisplay()}
            <AppForm onSubmit={handleSubmit}>
                <Form.Group>
                    <Form.Label>Name</Form.Label>
                    <Form.Control
                        autoFocus
                        id="name"
                        onChange={(e) => setName(e.target.value)}
                        size="lg"
                        value={name}
                    />
                </Form.Group>

                <Form.Group>
                    <Form.Label>Description</Form.Label>
                    <Form.Control
                        id="description"
                        onChange={(e) => setDescription(e.target.value)}
                        size="lg"
                        value={description}
                    />
                </Form.Group>

                <Button type="submit" disabled={isFormInvalid()}>
                    Add Group
                </Button>
            </AppForm>
            <AppError></AppError>
        </FormContainer>
    );
};
