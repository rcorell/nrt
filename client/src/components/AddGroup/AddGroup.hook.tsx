import { ApolloError } from '@apollo/client';
import { navigate } from 'hookrouter';
import { useState } from 'react';
import { useCreateGroupMutation } from 'src/api/__generated__/types';
import { Path } from '../Routes';

export const useAddGroup = () => {
    const [graphQLError, setGraphQLError] = useState(null as null | ApolloError);
    const [callMutation, { error }] = useCreateGroupMutation({
        onCompleted: () => {
            navigate(Path.GROUPS);
        },
        onError: setGraphQLError
    });

    const addGroup = (name: string, description: string) => {
        callMutation({
            variables: {
                description,
                name
            }
        });
    };

    const hook = {
        addGroup,
        error: error || graphQLError
    };

    return hook;
};
