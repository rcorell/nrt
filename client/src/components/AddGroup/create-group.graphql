mutation CreateGroup($name: String!, $description: String) {
    createGroup(name: $name, description: $description) {
        id
        createdAt
    }
}
