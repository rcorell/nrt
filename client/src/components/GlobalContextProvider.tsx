import React, { useState } from 'react';

import { isAuthenticated } from 'src/api/api';

type GlobalContextType = {
    authenticated: boolean;
    setAuthenticated: (authenticated: boolean) => void;
};

export const GlobalContext = React.createContext<GlobalContextType>({
    authenticated: false,
    setAuthenticated: (_authenticated) => {}
});

export const GlobalContextProvider: React.FC = ({ children }) => {
    const [authenticated, setAuthenticated] = useState(isAuthenticated());

    const contextFunctions = { authenticated, setAuthenticated };

    return <GlobalContext.Provider value={contextFunctions}>{children}</GlobalContext.Provider>;
};
