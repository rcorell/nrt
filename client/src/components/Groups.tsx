import React from 'react';
import * as _ from 'lodash';
import { useMutation, useQuery } from '@apollo/client';
import MaterialTable from 'material-table';

import { fetchGroupsQueryString, fetchUserQueryString, joinGroupMutationString } from 'src/api/api';
import { FetchGroupsQuery } from 'src/api/__generated__/FetchGroupsQuery';
import { FetchUserQuery } from 'src/api/__generated__/FetchUserQuery';
import { JoinGroupMutation, JoinGroupMutationVariables } from 'src/api/__generated__/JoinGroupMutation';
import { setBrowserTitle } from 'src/utils';

// const renderJoinColumn = (rowData: any) => {
//     if (userGroups.contains(rowData.name)) {
//         return <p>Joined</p>;
//     }

//     return <p>Join</p>;
// };

const COLUMN_DEFINITIONS = [
    { title: 'Name', field: 'name' },
    { title: 'Description', field: 'description' },
    {
        field: 'join',
        title: ''
    }
];

export const Groups: React.FC = () => {
    setBrowserTitle('Groups');

    const userQuery = useQuery<FetchUserQuery>(fetchUserQueryString);
    const allGroups = useQuery<FetchGroupsQuery>(fetchGroupsQueryString);

    const [joinGroup] = useMutation<JoinGroupMutation, JoinGroupMutationVariables>(joinGroupMutationString, {
        onCompleted: () => {
            // console.log('woo');
            userQuery.refetch();
        },
        onError: () => {
            // https://github.com/apollographql/apollo-client/issues/7167
        }
    });

    if (allGroups.loading || userQuery.loading) {
        return <div>Loading...</div>;
    }

    if (allGroups.error || userQuery.error) {
        return <div>Error: {JSON.stringify(allGroups.error) + JSON.stringify(userQuery.error)}</div>;
    }

    const userGroupIds = userQuery!.data!.user.groups.map((group) => {
        return group!.id;
    });

    const groups = _.cloneDeep(allGroups.data!.groups).map((group) => {
        return {
            ...group,
            join: userGroupIds.includes(group.id) ? (
                'Joined'
            ) : (
                <button onClick={() => joinGroup({ variables: { groupId: Number(group.id) } })}>Join!</button>
            )
        };
    });

    return <MaterialTable title="Groups" columns={COLUMN_DEFINITIONS} data={groups} />;
};
