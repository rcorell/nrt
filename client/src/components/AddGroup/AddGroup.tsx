import React, { useState } from 'react';
import { Button, Form, FormControlProps } from 'react-bootstrap';

import { AppError, AppForm, FormContainer } from 'src/styles/form';
import { setBrowserTitle } from 'src/utils';

import { useAddGroup } from './AddGroup.hook';

export const AddGroup: React.FC = () => {
    const [description, setDescription] = useState('');
    const [name, setName] = useState('');

    const addGroupHook = useAddGroup();

    const isFormInvalid = () => {
        return name.length < 1;
    };

    const handleSubmit = (event: React.FormEvent<FormControlProps>) => {
        event.preventDefault();

        addGroupHook.addGroup(name, description);
    };

    const errorDisplay = () => {
        if (addGroupHook.error) {
            return <div>Error: {JSON.stringify(addGroupHook.error)}</div>;
        }

        return null;
    };

    setBrowserTitle('Add Group');

    return (
        <FormContainer>
            <h1>Add Group</h1>
            {errorDisplay()}
            <AppForm onSubmit={handleSubmit}>
                <Form.Group>
                    <Form.Label id="nameLabel">Name</Form.Label>
                    <Form.Control
                        autoFocus
                        aria-labelledby="nameLabel"
                        id="name"
                        onChange={(e) => setName(e.target.value)}
                        size="lg"
                        value={name}
                    />
                </Form.Group>

                <Form.Group>
                    <Form.Label id="descriptionLabel">Description</Form.Label>
                    <Form.Control
                        id="description"
                        aria-labelledby="descriptionLabel"
                        onChange={(e) => setDescription(e.target.value)}
                        size="lg"
                        value={description}
                    />
                </Form.Group>

                <Button type="submit" disabled={isFormInvalid()}>
                    Add Group
                </Button>
            </AppForm>
            <AppError></AppError>
        </FormContainer>
    );
};
