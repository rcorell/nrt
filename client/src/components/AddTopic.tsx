import { useMutation, useQuery } from '@apollo/client';
import React, { useState } from 'react';
import { Button, Form, FormControlProps } from 'react-bootstrap';

import { CreateTopicMutation, CreateTopicMutationVariables } from 'src/api/__generated__/CreateTopicMutation';
import { FetchUserQuery } from 'src/api/__generated__/FetchUserQuery';
import { createTopicMutationString, fetchUserQueryString } from 'src/api/api';
import { LOADING_TEXT } from 'src/components/shared';
import { AppError, AppForm, FormContainer } from 'src/styles/form';
import { setBrowserTitle } from 'src/utils';

export const AddTopic: React.FC = () => {
    const [description, setDescription] = useState('');
    const [groupId, setGroupId] = useState('');
    const [title, setTitle] = useState('');

    const userQuery = useQuery<FetchUserQuery>(fetchUserQueryString);

    const [createTopic, { error }] = useMutation<CreateTopicMutation, CreateTopicMutationVariables>(
        createTopicMutationString,
        {
            variables: { groupId, title, description },
            onError: () => {
                // RTL bug
            }
        }
    );

    const isFormInvalid = () => {
        return title.length < 1;
    };

    const handleSubmit = (event: React.FormEvent<FormControlProps>) => {
        event.preventDefault();

        createTopic();
    };

    const errorStatus = () => {
        if (error) {
            return <div>Error: ${JSON.stringify(error)}</div>;
        }

        return null;
    };

    setBrowserTitle('Add Topic');

    if (userQuery.loading || !userQuery.data) {
        return <div>{LOADING_TEXT}</div>;
    }

    const groupOptions = userQuery.data.user.groups.map((group) => (
        <option key={group!.id} value={group!.id}>
            {group!.name}
        </option>
    ));

    return (
        <FormContainer>
            <h1>Add Topic</h1>
            {errorStatus()}
            <AppForm onSubmit={handleSubmit}>
                <Form.Group>
                    <Form.Label id="groupLabel">Group</Form.Label>
                    <Form.Control
                        as="select"
                        aria-labelledby="groupLabel"
                        id="group"
                        onChange={(e) => setGroupId(e.target.value)}
                        size="lg"
                    >
                        {groupOptions}
                    </Form.Control>
                </Form.Group>

                <Form.Group>
                    <Form.Label id="topicLabel">Topic</Form.Label>
                    <Form.Control
                        autoFocus
                        aria-labelledby="topicLabel"
                        id="topic"
                        onChange={(e) => setTitle(e.target.value)}
                        size="lg"
                        value={title}
                    />
                </Form.Group>

                <Form.Group>
                    <Form.Label id="descriptionLabel">Description</Form.Label>
                    <Form.Control
                        id="description"
                        aria-labelledby="descriptionLabel"
                        onChange={(e) => setDescription(e.target.value)}
                        size="lg"
                        value={description}
                    />
                </Form.Group>

                <Button type="submit" disabled={isFormInvalid()}>
                    Submit Topic
                </Button>
            </AppForm>
            <AppError></AppError>
        </FormContainer>
    );
};
