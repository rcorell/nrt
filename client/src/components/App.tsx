import 'bootstrap/dist/css/bootstrap.min.css';

import { ApolloProvider } from '@apollo/client';
import React from 'react';

import { GlobalContextProvider } from 'src/components/GlobalContextProvider';
import { apolloClient } from 'src/api/api';
import { Routes } from 'src/components/Routes';
import { AppHeader, GlobalStyle } from 'src/styles';

export const App: React.FC = () => {
    return (
        <>
            <GlobalStyle />
            <GlobalContextProvider>
                <ApolloProvider client={apolloClient}>
                    <AppHeader>Top 5 Daily</AppHeader>
                    <Routes />
                </ApolloProvider>
            </GlobalContextProvider>
        </>
    );
};
