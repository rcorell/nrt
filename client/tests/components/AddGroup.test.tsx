import { AddGroup } from 'src/components/AddGroup';

describe('AddGroup', () => {
    it('matches snapshot', () => {
        global.expectComponentToMatchSnapshot(AddGroup);
    });
});
