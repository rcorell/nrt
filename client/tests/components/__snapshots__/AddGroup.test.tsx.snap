// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddGroup matches snapshot 1`] = `
<div
  className="sc-bdvvaa hMSAon"
>
  <h1>
    Add Group
  </h1>
  <form
    className="sc-gsDJrp idIFaR"
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label
        className="form-label"
      >
        Name
      </label>
      <input
        autoFocus={true}
        className="form-control form-control-lg"
        id="name"
        onChange={[Function]}
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        className="form-label"
      >
        Description
      </label>
      <input
        className="form-control form-control-lg"
        id="description"
        onChange={[Function]}
        value=""
      />
    </div>
    <button
      className="btn btn-primary"
      disabled={true}
      type="submit"
    >
      Add Group
    </button>
  </form>
  <div
    className="sc-dkPtyc cWQCqc"
  />
</div>
`;
